elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java

```diff
@@ -0,0 +1,157 @@
+        Observable<PersistentResource> observableResults = exporter.export(queryObj);
+        Observable<String> downloadString = processObservablePersistentResource(observableResults);
+        storeResults(downloadString);
+
+        TableExportResult queryResult = new TableExportResult();
+        queryResult.setHttpStatus(200);
+        queryResult.setCompletedOn(new Date());
+        //TODO Add URL generation logic and set
+        //queryResult.setUrl(new URL("url"));
+        queryResult.setRecordCount(downloadRecordCount);
+        return queryResult;
+    }
+
+    /**
+     * This method stores the results.
+     * @param result is the observable result to store.
+     * @return TableExport object
+     */
+    protected TableExport storeResults(Observable<String> result) {
+        return resultStorageEngine.storeResults(queryObj, result);
+    }
+
+    /**
+     * Process Observable of Persistent Resource to generate the Observable of download-ready result string.
+     * @param resources Observable Persistent Resource.
+     * @return result as Observable of String
+     */
+    protected Observable<String> processObservablePersistentResource(Observable<PersistentResource> resources) {
+        Observable<String> results = Observable.just("No Records Generated");
+
+        if (queryObj.getResultType() == ResultType.CSV) {
+            results =  resources.map(resource -> {
+                incrementRecordCount();
+                return convertToCSV(resource);
+            });
+        } else if (queryObj.getResultType() == ResultType.JSON) {
+            results = resources.map(resource -> {
+                incrementRecordCount();
+                return resourceToJsonStr(resource);
+            });
+        }
+        return results;
+    }
+
+    protected String resourceToJsonStr(PersistentResource resource) throws IOException {
+        //TODO Additional logic to create an array of records???
+        return resource == null ? null : mapper.writeValueAsString(resource.getObject());
+    }
+
+    /**
+     * This method converts the JSON response to a CSV response type.
+     * @param resource is the Persistent Resource to convert
+     * @return returns string which is in CSV format
+     * @throws IllegalStateException Exception thrown
+     */
+    protected String convertToCSV(PersistentResource resource) {
+        if (resource == null) {
+            return null;
+        }
+
+        StringBuilder str = new StringBuilder();
+
+        List<Object[]> json2Csv;
+
+        try {
+            String jsonStr = resourceToJsonStr(resource);
+
+            JFlat flat = new JFlat(jsonStr);
+
+            json2Csv = flat.json2Sheet().headerSeparator("_").getJsonAsSheet();
+
+            int index = 0;
+
+            for (Object[] obj : json2Csv) {
+                // Skip Header record from 2nd time onwards.
+                // TODO Flag to Skip Headers
+                if (index++ == 0 && downloadRecordCount != 1) {
```

Comment:
aklish
Why do we check both index and downloadRecordCount. Isn't the latter sufficient?

 
moizarafat
convertToCSV is called once for each PersistentResource in the observable. json2Csv always has 2 entries. 0th index is the header so we need to skip the header from 2nd time this method is called. That's why we need to use a combination of index and downloadRecordCount.

Reply:
We need to add more comments here then.  Variables like index don't have enough description to understand that nuance.